packages/react-components/react-utilities/src/compose/slot.ts

```diff
@@ -0,0 +1,115 @@
+ * By default if `props.SLOT_NAME` is `undefined` then `state.SLOT_NAME` becomes `undefined`
+ * and nothing will be rendered, but if `required = true` then `state.SLOT_NAME` becomes an object
+ * with the values provided by `options.defaultProps` (or `{}`). This is useful for cases such as providing a default content
+ * in case no shorthand is provided, like the case of the `expandIcon` slot for the `AccordionHeader`
+ *
+ * @example of a required nullable slot
+ * ```tsx
+ * // AccordionHeader.types.ts
+ * type AccordionHeaderSlots = {
+ *    expandIcon?: Slot<'span'>
+ * }
+ * // useAccordionHeader.ts
+ * const state = {
+ *  expandIcon: slot(expandIconShorthand, {
+ *    required: true,
+ *    elementType: 'span',
+ *    defaultProps: { children: <ChevronRight/>, 'aria-hidden': true}
+ *  })
+ * }
+ * // renderAccordionHeader
+ * state.expandIcon && <state.expandIcon/>
+ * ```
+ */
+export function slot<Props extends UnknownSlotProps>(
+  value: Props | SlotComponent<Props> | SlotShorthandValue | undefined,
+  options: { required: true } & SlotOptions<Props>,
+): SlotComponent<Props>;
+export function slot<Props extends UnknownSlotProps>(
+  value: Props | SlotComponent<Props> | SlotShorthandValue | undefined | null,
+  options: { required?: boolean } & SlotOptions<Props>,
+): SlotComponent<Props> | undefined;
+export function slot<Props extends UnknownSlotProps>(
+  value: SlotComponent<Props>,
+  options?: { required: true } & Partial<SlotOptions<Props>>,
+): SlotComponent<Props>;
+export function slot<Props extends UnknownSlotProps>(
+  value: Props | SlotComponent<Props> | SlotShorthandValue | undefined | null,
+  options: { required?: boolean } & Partial<SlotOptions<Props>> = {},
+): SlotComponent<Props> | undefined {
+  const { required = false, defaultProps, elementType } = options;
+
+  if (value === null || (value === undefined && !required)) {
+    return undefined;
+  }
+
+  let metadata: SlotComponentMetadata<Props>;
+  if (isSlot<Props>(value)) {
+    metadata = value[SLOT_COMPONENT_METADATA_SYMBOL];
+    if (elementType !== undefined) {
+      metadata.elementType = elementType;
+    }
+  } else if (elementType !== undefined) {
+    metadata = { elementType };
+  } else if (process.env.NODE_ENV !== 'production') {
+    throw new Error("[react-utilities]: slot options.elementType is required when value isn't a slot itself");
+  } else {
+    metadata = { elementType: 'div' as React.ElementType<Props> as React.ComponentType<Props> };
+  }
+
+  /**
+   * Casting is required here as SlotComponent is a function, not an object.
+   * Although SlotComponent has a function signature, it is still just an object.
+   * This is required to make a slot callable (JSX compatible), this is the exact same approach
+   * that is used on `@types/react` components
+   */
+  const propsWithMetadata = {
+    ...defaultProps,
+    [SLOT_COMPONENT_METADATA_SYMBOL]: metadata,
+  } as SlotComponent<Props>;
```

Comment:
I made a PR against your local branch to show the changes for this suggestion :https://github.com/bsunderhus/fluentui/pull/1. It should work to just merge the PR, assuming you agree with this change. I think it's much cleaner to have a separation of concerns of the elementType vs. renderFunction, rather than combining them both on a metadata object. Thanks!

Reply:
> If you also take my suggestion of making elementType be a required argument to this slot function, then you could remove all of lines 73-85, and replace metadata.renderFunction on line 109 with propsWithMetadata[SLOT_RENDER_FUNCTION_SYMBOL].

It is required actually, there's only one case that `elementType` is optional, which is the case of you passing a slot  component to the `slot` method, in that case there's no reason to redeclare the `elementType` you can just take the one being provided from the slot component itself.

example:

```
const slotComponent = slot(shorthand, {required: true, elementType: 'div'})
// in this case here there's no reason to provide elementType again, just reuse the one from the slot component
const slotComponent2 = slot(slotComponent)
```

Reply:
But I'll look into simply using multiple symbols instead of metadata! seems legit as it would also avoid one deprecation

Reply:
> > If you also take my suggestion of making elementType be a required argument to this slot function, then you could remove all of lines 73-85, and replace metadata.renderFunction on line 109 with propsWithMetadata[SLOT_RENDER_FUNCTION_SYMBOL].
> 
> It is required actually, there's only one case that `elementType` is optional, which is the case of you passing a slot component to the `slot` method, in that case there's no reason to redeclare the `elementType` you can just take the one being provided from the slot component itself.

What's the use-case for this? I can't think of a reason you would want to redefine a slot after creating it. You might want to change the element type, but we could just provide a way to do that directly (e.g. by exporting `SLOT_ELEMENT_TYPE_SYMBOL`, or making a function like `modifySlotElementType`).

Reply:
I can't think of a good use case also TBH 🤷🏼‍♂️. I guess we can wait to see if that will be required, it would be possible to increment this without making breaking changes I believe